### ISSUE TITLE
IsFloat and IsDouble not consistant with GetFloat and GetDouble

### ISSUE THREAD

**Issue by Sumoren** [Tencent/rapidjson#1184]
If in the json a number value is a int like:
{
    "Number": 123
}
IsFloat(), Is<float>(), IsDouble() and Is<double>() return false while 123 is a valid float or double value.
Even GetFloat(), Get<float>(), GetDouble() and Get<double>() return the value correctly.

It limit what can be done in template since we cannot check Is<T>() to know if Get<T>() will return a valid value or assert without doing specialization for float and double to check IsNumber instead.

I would expect IsXXX to return true if it matching GetXXX can return a valid value.

**Comment by miloyip** [comment]
`GetDouble()` will be valid when `IsNumber() == true`.

Currently the API `IsDouble()` is for querying whether the value is stored in `double`. It does not handle implicit conversion in `GetDouble()`.

I think it may need template specialization for your need, which may specialized for a templated function, e.g. `IsConvertibleTo<double>(const Value& v) { return v.IsNumber(); }`.

**Comment by Sumoren** [comment]
From your answer I cannot even tell if you think the current behavior is correct or not.
Yes I can do a templated funciton with specialization for double/float and that what I did. But I would expect the API to better handle double.

About this comment:
>GetDouble() will be valid when IsNumber() == true.

Did you means in a future path? Because that not true. If you write this program:

    int main()
    {
        const char* json =
    R"({
        "Number": 123
    })";

        rapidjson::Document doc;
        doc.Parse(json);

        rapidjson::Value& number{ doc.FindMember("Number")->value };

        std::cout << std::boolalpha;
        std::cout << "IsNumber: " << number.IsNumber() << std::endl;
        std::cout << "Is<double>: " << number.Is<double>() << std::endl;
        std::cout << "IsDouble:" << number.IsDouble() << std::endl;

        return 0;
    }

You get this output:

    IsNumber: true
    Is<double>: false
    IsDouble:false

Also about your comment:
>It does not handle implicit conversion in GetDouble().

It look to me like GetDouble do implicit conversion from int, uint, int64 and uint 64:

    double GetDouble() const {
        RAPIDJSON_ASSERT(IsNumber());
        if ((data_.f.flags & kDoubleFlag) != 0)                return data_.n.d;   // exact type, no conversion.
        if ((data_.f.flags & kIntFlag) != 0)                   return data_.n.i.i; // int -> double
        if ((data_.f.flags & kUintFlag) != 0)                  return data_.n.u.u; // unsigned -> double
        if ((data_.f.flags & kInt64Flag) != 0)                 return static_cast<double>(data_.n.i64); // int64_t -> double (may lose precision)
        RAPIDJSON_ASSERT((data_.f.flags & kUint64Flag) != 0);  return static_cast<double>(data_.n.u64); // uint64_t -> double (may lose precision)
    }

**Comment by miloyip** [comment]
> It does not handle implicit conversion in `GetDouble()`.

Sorry I was intended to write:
> It does not handle implicit conversion **as** in `GetDouble()`.

I do not want to change the existing behavior as it is needed for some applications to differentiate `123.0` and `123`.

**Comment by Sumoren** [comment]
Ok thank you for the answer.

**Comment by mafranc** [comment]
The problem with this approach is that JSON does not distinguish between integers and floats, so JSON libraries are not obliged to add a ".0".

For example in Javascript:
`document.write(JSON.stringify(JSON.parse('{ "thisIsAFloat": 5.0, "thisIsAnInteger": 4}'))); `

will output
`{"thisIsAFloat":5,"thisIsAnInteger":4}`

In my case, I have an application that uses Qt to output a JSON which is then read by another application with rapidjson. The Qt JSON library does not write the ".0" when it is not necessary, and that is legitimate behavior. But it means I cannot directly use IsFloat to validate the value when parsing it.

In summary, I think it's incorrect for a JSON library to differentiate between 123.0 and 123, since in JSON the two numbers are equivalent. If the application needs this, it should add an extra flag, or expose the number as a string to be parsed by the application.
